src/pipeline.rs

```diff
@@ -17,30 +35,127 @@ impl Sink for Pipeline {
         &mut self,
         item: Self::SinkItem,
     ) -> Result<AsyncSink<Self::SinkItem>, Self::SinkError> {
-        self.inner.start_send(item)
+        // Note how this gets **swapped** with `new_working_set` in the loop.
+        // At the end of the loop, it will only contain finalized events.
+        let mut working_set = vec![item];
+        for inline in self.inlines.iter_mut() {
+            let mut new_working_set = Vec::with_capacity(working_set.len());
+            for event in working_set.drain(..) {
+                inline.transform(&mut new_working_set, event);
+            }
+            core::mem::swap(&mut new_working_set, &mut working_set);
+        }
+        self.enqueued.extend(working_set);
+
+        self.try_flush()?;
+        Ok(AsyncSink::Ready)
```

Comment:
If I'm reading this right it seems like we'll never return `NotReady` from `start_send`. I think we need to bound the size of `enqueued` and try to flush and `poll_complete` existing events before accepting and transforming more.

Reply:
Agree!

Reply:
I am not sure it is correct to cap the size of `enqueued` given that we know it will only be holding 1 event's work of processed messages at any given time and we are not permitted to block or discard messages here. Could you review https://github.com/timberio/vector/pull/4875/commits/4966d1fd5104b099b36e69654853042e2c1bbe07 see if this resolves your concerns?